src/internal/kopia/upload.go

```diff
@@ -722,7 +722,11 @@ func addMergeLocation(col data.BackupCollection, toMerge *mergeDetails) error {
 	prevLoc := lp.PreviousLocationPath()
 	newLoc := lp.LocationPath()
 
-	if prevLoc == nil {
+	if prevLoc == nil && newLoc == nil {
+		// assume this is a prefix-only movement, and does not require
+		// location information.
+		return nil
```

Comment:
> The type-check above should handle the situation you're talking about if you switch the collection type to something else

interface type casting is an unreliable measure of control here, since it cannot be asserted at compile time which can cascade to runtime gotchas.  The technicality that of being able to catch the scenario via that type check is not easily maintained in practice, and expects extra cognitive load from contributors.

> it becomes difficult to differentiate between a collection that was incorrectly created (a bug) and a known usecase (i.e. migration)

Hmm, I'm at odds here.  I think both approaches (interface type checking, input data case checking) are a smell, and have the potential for unexpected trouble.  Let me revisit the approach and see if I can find a middle ground.

Reply:
>The technicality that it can catch this conditions via the type check is not easily maintained in practice,

I think at that point it becomes more about enforcing patterns of usage than anything. The fact that we could use something with "empty" in the name (i.e. emptyCollection) sort of helps guide thinking that this is for a different thing than the other collections.

I would love to have compiler assertions, but I don't think we have an easy way to get that and nice code reuse, like we get by just sending a collection marked as moved to handle the migration, without adding another differently typed parameter

>expects extra cognitive load from contributors.

I think expecting folks to know which collections it's alright to return `(nil, nil)` location information from and which it isn't is also a cognitive load on contributors. Arguably one that's more difficult to understand since the current PR reuses a collection that has the same type as all the other collections returned. At least if the collection type is different it begs the question of why

Reply:
Forward note: we're generally in agreement here.  Just dithering on philosophy at this point.

> without adding another differently typed parameter

More or less what I plan on adding.  

> I think expecting folks to know which collections it's alright to return (nil, nil) location information from and which it isn't is also a cognitive load on contributors.

To be more specific, expecting contributors to know that one interface will be cast to another one under the hood requires extra cognitive load.  In terms of interface declaration, I'm a big proponent of "state expectations up front".  Ie: if a func is going to check an interface, the func should request that interface (or a super-set of it).  The current behavior hides that expectation within the code, which means the implementor won't know if they're supposed to provide compliance for that interface or not.  

Like you mentioned, using different structs to force thought about compliance is a good thing.  However, the complete picture does require implementors to easily connect the dots between their selection and its effects.

Reply:
>>without adding another differently typed parameter
>
>More or less what I plan on adding.

I'm not a big fan of adding another parameter to all of this. "Migration" is not something kopia wrapper really has any insight into. The bottom line is we're moving folders from point A to point B. I see no reason to further complicate things by adding yet another code path in what is already fairly complex code

>big proponent of "state expectations up front". Ie: if a func is going to check an interface, the func should request that interface (or a super-set of it). The current behavior hides that expectation within the code, which means the implementor won't know if they're supposed to provide compliance for that interface or not.

If I'm reading this right, the only way to really reach this condition would be to move both `LocationPath()` and `PreviousLocationPath()` to be part of the `BackupCollection` interface. We could do this, but we don't have the cycles to spare right now to accomplish that. Adding a separate "migration" parameter to work around the above further complicates the design because now we have folder moves and folder migrations, but migrations may only apply to some types of folder moves and the set of "some types of folder moves" is inconsistent

For example, what is happening with Exchange when we switch from UPN->UID is actually a migration. It just falls under the set of folder move operations that kopia wrapper can handle already and thus doesn't need a separate workaround. If we're trying to make separate structs for migrations so we can think of them as a separate sort of operation, then we'd somehow need to create an interface for kopia wrapper that would handle both the Exchange case and the OneDrive case.

Either way we end up with an unclear picture of what a "migration" is, though if we turn it into a parameter with it's own type it feels like we're trying to make it a special thing that we only actually use half the time while adding code complexity

>Like you mentioned, using different structs to force thought about compliance is a good thing. However, the complete picture does require implementors to easily connect the dots between their selection and its usage.

Attempting to get that in a complex code base can often cause more problems down the line than it's worth. As much as I hate to say it, we can really only push compile-time checking so far. Given that we've already made it rather difficult to generate "incorrect" paths and migrations are not expected to be a super common thing we need to do I'm willing to compromise on this and rely on just interface type-checks

Alternatively, you could see what happens if you give the migration collection an empty path builder for the current and previous locations. It will still comply to the interface, but should avoid the check. It will cause excess matches during details merging, but if both the old and new locations are empty strings in theory it won't harm anything. That feels a bit more like a hack than actually selecting only the behavior that's needed though

Reply:
Your mind is heading in a direction I wasn't planning to take the change.  I think it mediates appropriately between surfacing the consumers expectations while giving individual implementations an easy way out.